src/decorators/union.decorator.ts

```diff
@@ -0,0 +1,41 @@
+import { ApiProperty } from '@nestjs/swagger';
+
+import { IsBoolean, IsInteger, IsNumber, IsString } from '../flavors/openapi';
+
+export type OpenAPIDecoratorType = typeof IsBoolean
+    | typeof IsInteger
+    | typeof IsNumber
+    | typeof IsString;
+
+export type OpenAPIType = 'boolean' | 'integer' | 'number' | 'string';
+
+export interface OpenAPIRef {
+    $ref: string;
+}
+
+export function asType(item: OpenAPIDecoratorType): OpenAPIType {
+    switch (item) {
+        case IsBoolean:
+            return 'boolean';
+        case IsInteger:
+            return 'integer';
+        case IsNumber:
+            return 'number';
+        case IsString:
+            return 'string';
+        default:
+            break;
+    }
+
+    throw new Error('Missing case statement');
+}
+
+export function Union(types: OpenAPIDecoratorType[]): PropertyDecorator {
+    return ApiProperty({
+        oneOf: types.map(
+            (item) => ({
+                type: asType(item),
+            }),
+        ),
+    });
```

Comment:
There is one other possible approach that I can think of: we could allow Union if-and-only-if it used references to other types:
 - The array of values passed to `oneOf` can also use `{ $ref: ... }` format and the `getSchemaPath` function can help us resolve references at runtime.
 - There is a facility to register *extra* types with an OpenAPI schema. I think it's possible to invoke the `ApiExtraModels` decorator from within `Union`.

The resulting types would look like:

```ts
interface FooBarNumber {
   value: string;
}
interface FooBarString {
   value: string;
}
interface Foo {
   bar: FooBarNumber | FooBarString;
}
```

And the OpenAPI:

```json
  "components": {
    "schemas": {
      "FooBarNumber": {
           ...
       },
      "FooBarString": {
           ...
       },
      "Foo": {
        "properties": {
          "bar": {
            "oneOf": [
              { "$ref": "#/components/schemas/FooBarNumber" }
              { "$ref": "#/components/schemas/FooBarString" }
            ],
          },
        },
        "required": [
          "bar",
        ],
        "type": "object",
      },
    },
  },
```

We lose slightly with our expressiveness (e.g. requiring `.value`), but end up with something a lot more composable.